### PR TITLE
feat: integrate REVM instead of SputnikEVM

### PR DESCRIPTION
**Pull request by gianbelinche**
# What ❔
This PR adds integration with [REVM](https://github.com/bluealloy/revm)

To run the semantic tests:
`cargo run --verbose --release --bin compiler-tester -- --path solidity/test/libsolidity/semanticTests --mode 'Y+ 0.8.26' --target EVM --use-upstream-solc`

These are the differences on Semantics Tests with the EVM Interpreter

```
Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[addr:24] was failing and is now passing ✅
Test solidity/test/libsolidity/semanticTests/array/array_storage_index_access.sol[test_indices:7] was failing and is now passing ✅
Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[owner:22] was failing and is now passing ✅
Test solidity/test/libsolidity/semanticTests/functionCall/precompile_extcodesize_check.sol[testLowLevel:2] was failing and is now passing ✅
Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[disown:20] was failing and is now passing ✅
Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[subRegistrar:25] was failing and is now passing ✅
Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[content:23] was failing and is now passing ✅
Test solidity/test/libsolidity/semanticTests/tryCatch/create.sol[g:2] was passing and is now failing ❌
Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[#balance_check:21] was passing and is now failing ❌
```

Here is an explanation of the differences

> Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[owner:22] was failing and is now passing ✅
> Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[content:23] was failing and is now passing ✅
> Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[subRegistrar:25] was failing and is now passing ✅
> Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[addr:24] was failing and is now passing ✅
> Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[disown:20] was failing and is now passing ✅
> Test solidity/test/libsolidity/semanticTests/externalContracts/FixedFeeRegistrar.sol[#balance_check:21] was passing and is now failing ❌

Since the address 0x124 is not a system contract on revm, the send can be done correctly, however, this made us realize that the test is wrong, since the balance_check:21 should actually be 69 eth, which is the correct value revm outputs, it does not fail on evm interpreter since the disown does not work.

> Test solidity/test/libsolidity/semanticTests/functionCall/precompile_extcodesize_check.sol[testLowLevel:2] was failing and is now passing ✅

Revm has the identity precompile, while evm interpreter does not.

> Test solidity/test/libsolidity/semanticTests/array/array_storage_index_access.sol[test_indices:7] was failing and is now passing ✅

On EVM Interpreter we don't have enough gas (0xffffff), we do however on revm (20000000)

> Test
solidity/test/libsolidity/semanticTests/tryCatch/create.sol[g:2] was passing and is now failing ❌

In this test, first it tries to deploy a contract which constructor reverts, it catches that error and then it passes.
Then it deploys another contract, this one does not revert, however the address expected is not the same.
This is because on EVM when a deploy fails, the nonce is still incremented, and this test expects it is not.

For the remaining tests this are the numbers:

REVM:
16673 Success
400 Failed
2 Invalid

EVM Interpreter:
16802 Success
269 Failed
4 Invalid

## Checklist

- [x] PR title corresponds to the body of PR.
- [x] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `cargo fmt` and checked with `cargo clippy`.

### CODE REVIEW

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Thanks a lot for this integration!
Please note that I've renamed EVM to EVMEmulator in the base branch, and refactored some code.
Could you rebase?

**Review by hedgar2017 — APPROVED** [review]

File: compiler_tester/src/test/case/input/deploy_evm.rs [hunk 1]
```diff
@@ -110,6 +131,116 @@ impl DeployEVM {
...
+                env.tx.transact_to = TxKind::Create;
+            })
+            .build();
+
+        let res = match new_vm.transact_commit() {
+            Ok(res) => res,
+            Err(error) => {
+                match error {
+                    EVMError::Transaction(e) => {
+                        println!("Error on transaction: {:?}", e)
```
Comment (hedgar2017) [inline compiler_tester/src/test/case/input/deploy_evm.rs:187]: We are not printing anything to std.
Please pass the error to `Summary::invalid` below instead.

File: compiler_tester/src/test/case/input/runtime.rs [hunk 2]
```diff
@@ -165,6 +187,121 @@ impl Runtime {
...
+                env.tx.access_list = vec![];
+            })
+            .build();
+
+        vm = self.update_balance_if_lack_of_funds(caller, vm);
+
+        let res = match vm.transact_commit() {
+            Ok(res) => res,
+            Err(error) => {
+                match error {
```
Comment (hedgar2017) [inline compiler_tester/src/test/case/input/runtime.rs:245]: Please pass the error to `Summary::invalid`.

File: compiler_tester/src/vm/eravm/system_context.rs [hunk 3]
```diff
@@ -97,6 +114,54 @@ impl SystemContext {
     const ZERO_BLOCK_HASH_EVM: &'static str =
         "0x3737373737373737373737373737373737373737373737373737373737373737";
 
+    pub const L2_ETH_TOKEN_ADDRESS: Address = H160([
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x80, 0x0a,
+    ]);
+
+    pub fn address_to_h256(address: &Address) -> H256 {
```
Comment (hedgar2017) [inline compiler_tester/src/vm/eravm/system_context.rs:122]: Please move to `crate::utils`, where other such functions reside.

File: compiler_tester/src/vm/eravm/system_context.rs [hunk 4]
```diff
@@ -97,6 +114,54 @@ impl SystemContext {
     const ZERO_BLOCK_HASH_EVM: &'static str =
         "0x3737373737373737373737373737373737373737373737373737373737373737";
 
+    pub const L2_ETH_TOKEN_ADDRESS: Address = H160([
```
Comment (hedgar2017) [inline compiler_tester/src/vm/eravm/system_context.rs:117]: Please use the constant from `zkevm_opcode_defs`, like I did in the base PR.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Looks good!
Just some nits.

**Comment by hedgar2017** [comment]
Please also run `cargo fmt` pre-commit.

File: compiler_tester/src/test/case/input/deploy_evm.rs [hunk 5]
```diff
@@ -184,22 +184,46 @@ impl DeployEVM {
             Err(error) => {
                 match error {
```
Comment (hedgar2017) [inline compiler_tester/src/test/case/input/deploy_evm.rs:185]: Can't this match be avoided by just passing the whole error to the function?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

File: compiler_tester/src/test/case/input/deploy_evm.rs (hunk 5 above)
Comment (gianbelinche) [reply compiler_tester/src/test/case/input/deploy_evm.rs:185]: The thing is EVMError::Database does not implement Display, so in order to pass it to summary I need to match it.

Comment (hedgar2017) [reply compiler_tester/src/test/case/input/deploy_evm.rs:185]: OK, let's keep it for now.
Thanks!

**Review by hedgar2017 — APPROVED** [review]

**Pull request merged**